kubespawner/proxy.py

```diff
@@ -0,0 +1,244 @@
+import os
+import string
+import escapism
+import json
+from kubernetes import client
+
+from jupyterhub.proxy import Proxy
+from jupyterhub.utils import exponential_backoff
+
+from kubespawner.objects import make_ingress
+from kubespawner.utils import generate_hashed_slug
+from kubespawner.reflector import NamespacedResourceReflector
+from concurrent.futures import ThreadPoolExecutor
+from traitlets import Unicode
+from tornado import gen
+from tornado.concurrent import run_on_executor
+
+
+class IngressReflector(NamespacedResourceReflector):
+    kind = 'ingresses'
+    labels = {
+        'heritage': 'jupyterhub',
+        'component': 'singleuser-server',
+        'hub.jupyter.org/proxy-route': 'true'
+    }
+
+    list_method_name = 'list_namespaced_ingress'
+    api_group_name = 'ExtensionsV1beta1Api'
+
+    @property
+    def ingresses(self):
+        return self.resources
+
+class ServiceReflector(NamespacedResourceReflector):
+    kind = 'services'
+    labels = {
+        'heritage': 'jupyterhub',
+        'component': 'singleuser-server',
+        'hub.jupyter.org/proxy-route': 'true'
+    }
+
+    list_method_name = 'list_namespaced_service'
+
+    @property
+    def services(self):
+        return self.resources
+
+class EndpointsReflector(NamespacedResourceReflector):
+    kind = 'endpoints'
+    labels = {
+        'heritage': 'jupyterhub',
+        'component': 'singleuser-server',
+        'hub.jupyter.org/proxy-route': 'true'
+    }
+
+    list_method_name = 'list_namespaced_endpoints'
+
+    @property
+    def endpoints(self):
+        return self.resources
+
+class KubeIngressProxy(Proxy):
+    namespace = Unicode(
+        config=True,
+        help="""
+        Kubernetes namespace to spawn ingresses for single-user servers in.
+
+        If running inside a kubernetes cluster with service accounts enabled,
+        defaults to the current namespace. If not, defaults to 'default'
+        """
+    )
+
+    def _namespace_default(self):
+        """
+        Set namespace default to current namespace if running in a k8s cluster
+
+        If not in a k8s cluster with service accounts enabled, default to
+        'default'
+        """
+        ns_path = '/var/run/secrets/kubernetes.io/serviceaccount/namespace'
+        if os.path.exists(ns_path):
+            with open(ns_path) as f:
+                return f.read().strip()
+        return 'default'
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        # We want 3x concurrent spawn limit as our threadpool. This means that if
```

Comment:
How do you feel about this logic, @minrk? It vaguely feels right to me, but I can't make a strong case for it yet.

Reply:
We never actually will have all 3 of these requests going at once, since we wait for each object to show up in our reflector first. Maybe we don't have to do that (and can just rely on the redirects=n looping to fix it), since that will give us a big performance boost.

Perhaps the better thing to do is to figure out how to put multiple objects atomically in the same http request. That seems like the 'right' thing to do.

Reply:
Maybe use 1x instead of 3x? It seems okay when hitting capacity to allow a single spawn's requests to serialize. It doesn't seem important that all possible requests are in flight at once.

Reply:
I agree with @minrk on 1x. It will be simpler to test initially and can be increased in a future refactor if needed.

Reply:
Yeah, I agree! Upon further thought, this will also never use more than 1x, since we serialize these creates anyway. I'll update!